### PR TITLE
Jaimeta/newfolder

### PR DESCRIPTION
**Pull request by jtarquino**
Fixes # .

Changes proposed in this pull request:
 - added a new cmdlet to create folders in the server

### CODE REVIEW

**Comment by FriedrichWeinmann** [comment]
Hi Jaime,

some issues with your function:

- Parameter help is broken (Adding '(optional)' to the name breaks help for parameters)
- Don't use break to terminate a function: It can't be caught and will rip through anything above (other functions/script calling this function) until it finds a loop it can break. Instead throw an exception and add the caught exception as a nested exception: That way the user can drill down in the error and find the original one. It also becomes possible to try/catch the function.
My Pull request contains a lot of examples for this, in case you want to know how that is implemented (in fact, it's one of the main reasons I want and wrote that request)
- Not really an issue, but you may want to reorder your parameters: The parameter should be in descending order of importance to the user: When you use tab to browse the available parameters, they are presented in just that order.

**Comment by SQLvariant** [comment]
I might suggest a more obvious parameter name than just -Name.  Maybe -Folder, -NewFolder, or even in this case -FolderName.  

Using -FolderName would be my least favorite of the 3, but in this case, since the folder parameter in question is not some sort of SMO-Type, it's just a String, I would be ok with that parameter name.

I'd love to hear what other folks think.

**Comment by SQLvariant** [comment]
It seems like your -Destination parameter might be equivalent to the -Path parameter used in the commands below.  If so, I would recommend using -Path for consistency.

- Get-RsCatalogItems   
- Out-RsCatalogItem    
- Out-RsFolderContent  
- Remove-RsCatalogItem 
- Write-RsCatalogItem  
- Write-RsFolderContent

**Comment by jtarquino** [comment]
@FriedrichWeinmann great catch with the optional breaking the help , all your comments are very helpful.
It will take a little while for the team to review your PR due the size but I'll keep you posted

**Comment by jtarquino** [comment]
@SQLvariant lets go with FolderName and Path , those seem to be the most explicit and bullet proof names

**Comment by FriedrichWeinmann** [comment]
Hi Jaime,

don't worry about the verification taking a while: I _did_ dump quite a big lump at you in a single go. Not much of a choice given some of the changes I did implement though (especially the connection management).

Regarding parameter naming and conventions:
 - If you want to tie into my changes: Add an alias to 'Path' named 'ItemPath'. In the original 0.0.8 both were alternatingly present, my update makes it uniform in that every one of them is named 'Path' and has the alias 'ItemPath', so both will work.
 - I'd go with 'Name', rather than 'FolderName', since that's the general practice in most of PowerShell. You could keep 'FolderName' on as an alias, for those that find that more intuitive, but common practice says 'Name'
 - The default name for a parameter used to authenticate a connection over is 'Credential'. I thus renamed all instances of 'ReportServerCredentials' and kept the old name as an alias to keep from wantonly breaking stuff. Also one of the general best practices of coding style is to keep function- and parameter names singular, not plural.
 - Your current way of error handling will produce two errors, one of which cannot easily be caught. Rather than doing it like this:
`Write-Error "Exception occurred while creating folder! $($_.Exception.Message)"`
`throw`
You might want to throw a new exception object:
`throw (New-Object System.Exception("Exception occurred while creating folder! $($_.Exception.Message)", $_.Exception))`

**Comment by jtarquino** [comment]
I'm aligning with your changes to minimize friction during the integration with your PR, ( the alias option is a neat trick)
Regarding the exceptions I will trust you in this one and the Power Shell expectations / usage, although typically we don't use exceptions to control the flow of the code (a classic example is TryParse versus Convert.ToInt).

Looking forward to check your changes looks like they will very valuable for the project.

**Comment by SQLvariant** [comment]
@FriedrichWeinmann the rest of PowerShell may have a lot of -Name parameters but SQL Server and it's various tools have rich defined object models (e.g. SMO) and the large user base of SQL Server is used to referring to objects by their type (but not their specific SMO type, which is much longer).  While "Name" is frequently found throughout the SMO I find that the users of SQL Server find the object names much more intuitive.  For example the Get-SqlDatabase cmdlet has a -Name parameter that people overlook.  On the other hand the Backup-SqlDatabase cmdlet has a -Database parameter and people know exactly what to do with that.  Furthermore, command discoverability is enhanced when someone can simply search for Get-Command -Module SqlServer -ParameterName Database to find a list of cmdlets that allow them to work with databases, as opposed to Get-Command -Module SqlServer -ParameterName Name which really doesn't tell you much of anything.  As it stands today, half of the cmdlets found when running Get-Command -Module SqlServer -ParameterName Database do not have the word Database anywhere in the noun portion of their name.  

I'm ok with a parameter being aliased to Name, but I'm not ok with calling the parameter Name, if it can be avoided.

I hope this explains why I don't find -Name to be a helpful parameter name?

**Comment by SQLvariant** [comment]
@jtarquino Please seem my thoughts above on why using -Name is detrimental to command discoverability.  I am open to aliasing a parameter name to 'Name' but I find many expect a more descriptive indication of what the parameter accepts.  And after all, that's what PowerShell is all about, making things discoverable, not vague and confusing.

File: Functions/CatalogItems/New-RsFolder.ps1 [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        [Parameter(Mandatory=$True)]
+        [Alias('FolderName')]
+        [string]
+        $Name,
+
+        [string]
+        $ReportServerUri = 'http://localhost/reportserver',
+
+        [System.Management.Automation.PSCredential]
+        $Credential,
```
Comment (parthsha) [inline Functions/CatalogItems/New-RsFolder.ps1]: Rename this to $ReportServerCredentials to be consistent with other scripts.

**Comment by jtarquino** [comment]
After hearing both sides and the great feedback I have decided to err in side of the extra discoverability and friendliness of the cmdlet.

I'll also use the old name for the credentials until we move forward with the big refactor to keep consistency with the existing naming and then after that I'll update the function so the changes are small and localized.

I really appreciate your passion and contribution , this set of discussions are exactly what we were looking when we started the project

**Comment by FriedrichWeinmann** [comment]
@jtarquino : Regarding the throwing of exceptions as well as try/catch: In PowerShell, quite a few debugging mechanics and flow control options tie into the practice of throwing and catching errors. For example I can use the generic "-ErrorAction" parameter to determine, whether a function that throws an error crashes my own function around my ears ... or whether I'll ignore the error completely! (In fact, the '[CmdletBinding()]' attribute above the param blocks adds this (and other) parameter to your function). When running a command manually, I can check out the exception thrown, inspect inner exceptions and so on, so I can easily figure out manually, just why something failed and exactly where it did do so. This is also why it's important to throw exception objects with inner exceptions when you have one available (and why it's enough to do a `throw "foo"` when one hasn't just caught an exception to pass along).

@SQLvariant : I totally hear you. I'm coming very much from the PS side of things and only occasionally stray into SQL territory, which makes me stronger with PowerShell standards and less well versed with SQL community expectations. A little info regarding your point: Get-Command will also find commands by parameter alias, just as it would find a command by the parameters 'real' name. The only real difference is that when you tab through possible parameters, you will only tab through the original names, not the aliases (unless you already have part of the parameter name typed before wielding tab, then it will also find aliases). So by all means, keep the name the majority of the user-base will find more intuitive as the main name. In fact, wouldn't be a good idea to define naming conventions in a comments file? That would provide a central reference as well as offering a basis for discussion.

**Pull request merged**